scarlet/source.py

```diff
@@ -132,56 +121,46 @@ def build_detection_coadd(sed, bg_rms, observation, thresh=1):
+    bg_cutoff = np.sqrt((weights ** 2 * positive_bgrms ** 2).sum()) / jacobian
     return detect, bg_cutoff
 
 
-def init_extended_source(sky_coord, frame, observation, bg_rms,
-                         thresh=1., symmetric=True, monotonic=True):
-    """Initialize the source that is symmetric and monotonic
-    See `ExtendedSource` for a description of the parameters
-    """
-    # determine initial SED from peak position, accounting for PSF changes
-    sed = get_psf_sed(sky_coord, observation, frame)
-
-    if np.any(sed <= 0):
-        # If the flux in all channels is  <=0,
-        # the new sed will be filled with NaN values,
-        # which will cause the code to crash later
-        msg = "Zero or negative SED {} at y={}, x={}".format(sed, *sky_coord)
-        if np.all(sed <= 0):
-            logger.warning(msg)
-        else:
-            logger.info(msg)
-
-    morph, bg_cutoff = build_detection_coadd(sed, bg_rms, observation, thresh)
-    center = frame.get_pixel(sky_coord)
-
-    # Apply the necessary constraints
-    if symmetric:
-        morph = operator.prox_uncentered_symmetry(morph, 0, center=center, algorithm="sdss")
-    if monotonic:
-        # use finite thresh to remove flat bridges
-        prox_monotonic = operator.prox_strict_monotonic(morph.shape, use_nearest=False,
-                                                        center=center, thresh=.1)
-        morph = prox_monotonic(morph, 0).reshape(morph.shape)
-
+def trim_morphology(sky_coord, frame, morph, bg_cutoff, thresh):
     # trim morph to pixels above threshold
-    mask = morph > bg_cutoff
-    if mask.sum() == 0:
-        msg = "No flux above threshold={2} for source at y={0} x={1}"
-        raise SourceInitError(msg.format(*sky_coord, bg_cutoff))
-    morph[~mask] = 0
-
-    # normalize to unity at peak pixel
-    cy, cx = np.array(center).astype(int)
-    center_morph = morph[cy, cx]
-    morph /= center_morph
-    return sed, morph
-
-
-def init_combined_extended_source(sky_coord, frame, observations, bg_rms, obs_idx=0,
-                                  thresh=1., symmetric=True, monotonic=True):
+    mask = morph > bg_cutoff * thresh
+    boxsize = 16
+    pixel_center = frame.get_pixel(sky_coord)
+    if mask.sum() > 0:
+        morph[~mask] = 0
+
+        # normalize to unity at peak pixel
+        center_morph = morph[pixel_center[0], pixel_center[1]]
+        morph /= center_morph
+
+        # find fitting bbox
+        bbox = Box.from_data(morph, min_value=0)
+        boxsize = 16
+        if bbox.contains(pixel_center):
+            size = 2 * max((pixel_center[0] - bbox.bottom, bbox.top - pixel_center[0],
+                            pixel_center[1] - bbox.left, bbox.right - pixel_center[1]))
+            while boxsize < size:
+                boxsize *= 2
+    else:
+        msg = "No flux above threshold for source at y={0} x={1}".format(*center)
+        logger.warning(msg)
```

Comment:
We mangle two arguments here:
* Can I initialize a source this way?
* Should I use a different source?

The second is much more complicated to address. It should not be done at this stage. A different way of saying this is: I want to see the fitter fail, not the init. Your recent tests showed that we fail mostly because of the init, which means that the fitter didn't even get a chance to try. That is inconsistent behavior. In addition, we now allow source models that can be fit under different constraints (e.g. dust, which is negative potentially in all observed bands). A warning is justified, but we cannot have it throw an error.

Reply:
I think I should be more clear. The initialization already fails and throws an error, but it's an obscure numpy error due to `nan` values that are generated when using an invalid SED. The old stack wrapper for scarlet could catch `SourceInitError` exceptions since we expected those, but would still throw an error if something went wrong that we didn't expect.

Something changed and now I have to catch _everything_, otherwise 20-30% of the HSC blends fail during the init. So that means that there are other issues with the initialization and instead of just letting the code fail, when we can identify the source of obscure errors we should do just that and throw our own exception that makes it clear to the user what happened.

But this is the one place where the `adam` branch is actually much worse! Sometimes it throws a numpy exception, which I can catch if I catch all errors (which is a bad programming practice) but for multi-component sources somehow it in initializing with a bunch of `nan` values. Then when it tries to fit the model the system hangs, which is even worse than failing! For a user looking a single blend a warning might be find but to run scarlet at scale, even on a few arcminutes of data, we have to have some way of notifying the user that blending failed, either by throwing an error or by setting a flag that they can read. This is critical for using scarlet in LSST and the pipeline will not work without it.

Reply:
I get some of that, but we need to be very specific about the language here because I still think we talk somewhat past each other.

* `nan` arise from normalizing a zero vector. This can arise if you combine positivity with some form of normalization.  just having positivity is not a problem, this is why the exception should not be thrown there, because there can well be components that are fine with it. In essence, source should check their inits because they know the constraints. Doing so should suffice for LSST.
* "when it tries to fit the model the system hangs, which is even worse than failing!" Please clarify what "hangs" means here. I'm guessing that once a single nan arises, numpy switches to survival mode (dealing with warnings/exceptions internally), which slows it down. If the initialization is fine and the constraints are feasible, this should never happen. However, even with proper init, the constraints may not be mutually feasible, and we should probably think of a way to fail gracefully then.
* "we have to have some way of notifying the user that blending failed". I don't think that blending failed, it's something about the modeling that failed. See item above.

So, what I suggest is having two different kinds of exceptions: `SourceInitError` and `FittingError`. The former is source-specific, the latter is some form of runtime error that probably is generic/global. It could for instance be a check of `nan` or `inf` in the parameters. I'll add some code to that effect. But to be clear, this check will cost runtime.

Reply:
On second thought, maybe going generic is the right approach here. `ParameterError` could indicate that some elements are improper. The code for checking is a `Parameter` method and will be called once by `Component.__init__` and by `Blend.fit` after every another gradient step.

Reply:
Ok, this looks good as long as the effect on runtime is negligible. I checked and it actually looks like (in a few blends that I just looked at) when a source is improperly initialized it has `nan` values from the initialization before fitting even begins. So it might be simple enough to just check the initial SED and morphology for `nan` values if this affects the performance too much.